pkg/operator/controllers/autosizednodes/autosizednodes_controller_test.go

```diff
@@ -0,0 +1,145 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+
+	// This "_" import is counterintuitive but is required to initialize the scheme
+	// ARO unfortunately relies on implicit import and its side effect for this
+	arov1alpha1 "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
+	_ "github.com/Azure/ARO-RP/pkg/util/scheme"
+)
+
+func TestAutosizednodesReconciler(t *testing.T) {
+	aroMeta := metav1.ObjectMeta{
+		Name:      "aro",
+		Namespace: "openshift-azure-operator",
+	}
+
+	emptyConfig := mcv1.KubeletConfig{}
+	config := makeConfig()
+
+	tests := []struct {
+		name       string
+		wantResult ctrl.Result
+		wantErr    error
+		wantGetErr error
+		client     client.Client
+		wantConfig *mcv1.KubeletConfig
+	}{
+		{
+			name:       "is not needed",
+			wantResult: ctrl.Result{},
+			client: fake.NewClientBuilder().WithRuntimeObjects(
+				&arov1alpha1.Cluster{
+					ObjectMeta: aroMeta,
+					Spec: arov1alpha1.ClusterSpec{
+						Features: arov1alpha1.FeaturesSpec{
+							ReconcileAutoSizedNodes: false,
+						},
+					},
+				},
+			).Build(),
+			wantConfig: &emptyConfig,
+			wantErr:    nil,
+			wantGetErr: kerrors.NewNotFound(mcv1.Resource("kubeletconfigs"), "dynamic-node"),
+		},
+		{
+			name:       "is needed and not present already",
+			wantResult: ctrl.Result{},
+			client: fake.NewClientBuilder().WithRuntimeObjects(
+				&arov1alpha1.Cluster{
+					ObjectMeta: aroMeta,
+					Spec: arov1alpha1.ClusterSpec{
+						Features: arov1alpha1.FeaturesSpec{
+							ReconcileAutoSizedNodes: true,
+						},
+					},
+				}).Build(),
+			wantConfig: &config,
+			wantErr:    nil,
+			wantGetErr: nil,
+		},
+		{
+			name:       "is needed and present already",
+			wantResult: ctrl.Result{},
+			wantErr:    nil,
+			client: fake.NewClientBuilder().WithRuntimeObjects(
+				&arov1alpha1.Cluster{
+					ObjectMeta: aroMeta,
+					Spec: arov1alpha1.ClusterSpec{
+						Features: arov1alpha1.FeaturesSpec{
+							ReconcileAutoSizedNodes: true,
+						},
+					},
+				},
+				&config).Build(),
+			wantConfig: &config,
+		},
+		{
+			name:       "is not needed and is present",
+			wantResult: ctrl.Result{},
+			client: fake.NewClientBuilder().WithRuntimeObjects(
+				&arov1alpha1.Cluster{
+					ObjectMeta: aroMeta,
+					Spec: arov1alpha1.ClusterSpec{
+						Features: arov1alpha1.FeaturesSpec{
+							ReconcileAutoSizedNodes: false,
+						},
+					},
+				},
+				&config).Build(),
+			wantConfig: &emptyConfig,
+			wantErr:    nil,
+			wantGetErr: kerrors.NewNotFound(mcv1.Resource("kubeletconfigs"), "dynamic-node"),
+		},
+	}
+
+	cluster := arov1alpha1.Cluster{}
+	cluster.DeepCopyObject()
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			ctx := context.Background()
+			r := Reconciler{
+				Client: test.client,
```

Comment:
Nit: I would probably replace `test.client` with `test.reconcileAutoSizedNodes` which is either `true` or `false`. Should make slice of `tests` a bit less verbose.

```suggestion
				Client: fake.NewClientBuilder().WithRuntimeObjects(
					&arov1alpha1.Cluster{
						ObjectMeta: aroMeta,
						Spec: arov1alpha1.ClusterSpec{
							Features: arov1alpha1.FeaturesSpec{
								ReconcileAutoSizedNodes: test.reconcileAutoSizedNodes,
							},
						},
					},
					&config).Build(),
```